File: ANTEMA_main.m
% Main script for Segmentation Analysis of HRTEM images
% Run this script to run the automated nanoparticle transmission electron
% micrograph analyzer (ANTEMA)


% Copyright (C) 2022  University of Duisburg-Essen, 
%                       Nina Gumbiowski
% 
%     This program is free software: you can redistribute it and/or modify
%     it under the terms of the GNU General Public License as published by
%     the Free Software Foundation, either version 3 of the License, or
%     (at your option) any later version.
% 
%     This program is distributed in the hope that it will be useful,
%     but WITHOUT ANY WARRANTY; without even the implied warranty of
%     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%     GNU General Public License for more details.
% 
%     You should have received a copy of the GNU General Public License
%     along with this program.  If not, see <http://www.gnu.org/licenses/>.


clear
close all
addpath('SegmentationFunctions','Nets','dm3read')

%% Settings

netname = 'NP-net-weights.mat';
Convexthresh = 0.95;    % threshold for particle separation based on convexity
PSep = true;            % Particle separation enabled (true) or disabled (false)
saveas = 'xlsx';        % set to 'csv' or 'xlsx'
removecount = 0.5;      % minimum particle size in nm
minmarker = 0.5;        % minimum marker size for particle separtion routine in nm
Mode = 'UECS';          % Mode : Mode for particle separation. Can either be 'UECS' or 'Watershed'
                        %           Default: UECS
% Colormap for the raw segmentation map
cmap = [
         0 114 189      % Background
         237 177 32     % Particle 
         ];
     cmap = cmap ./ 255;
%% load network for segmentation
load(netname, 'net')

%% Initialize as batch or single file
batch=input('Segment Batch or single file?\n Press:\n 0 : single file\n 1 : batch\n');
if batch==0 % single file option
    fprintf('Choose dm3 or tif file to be segmented \n')
    [baseName, pathdir] = uigetfile({'*.dm3;*.tif'});
    filename = {fullfile(pathdir, baseName)};
       
elseif batch==1 % batch option for all images in one folder
    fprintf('Choose folder that contains all .dm3 or .tif files to be segmented \n')
    pathdir=uigetdir();
    fileformat = input('What file format is your data in? \nPress: \n1) .tif \n2) .dm3 \n');
    if fileformat == 1
        [stat,scan]=fileattrib([pathdir '\*.tif']);
    else 
        [stat,scan]=fileattrib([pathdir '\*.dm3']);
    end
    filename={scan.Name};
    fprintf('%d files found in specified folder\n',length(scan)) 
    allinone = input('How should the data be saved?\n 0: one file for each image\n 1: all image data in one file\n 2: both\n'); %%IMPLEMENT SAVING OPTION AT THE END
    if batch == 1  && allinone > 0
        AllpartProp = [];
    end
else
    fprintf('ERROR: Irregular input for Segment Batch or single file\n')
    return
end

%% start timer for performance evaluation
tic

%% Otion to save elsewhere
store = input('Where should the results be stored?\n 1: Results folder (default)\n 2: Directory of the origin data\n 3: Other \n');
if store == 1
    resfolder = 'Results';
elseif store == 2
    resfolder = pathdir;
elseif store == 3
    resfolder = uigetdir();
else
    fprintf('ERROR: Irregular input \n')
    return
end
%% Processing images
for i=1:length(filename)
    close all
    % load singular dm3 file / tif file
    filepath=filename{i};
    [pathstr,expName,ext] = fileparts(filepath);
    if strcmp(ext,'.dm3')==true
        [Image,pxsz,units]=ReadDMFile(filepath);
        if isempty(units)
            warning('Unit of dm3 file is empty, Skip evaluation of %s\n',filepath)
            continue
         elseif units~='nm'
            warning('Unit not in nm, Case currently not implemented\n Skip evaluation of %s\n',filepath)
            continue
        end
    elseif strcmp(ext,'.tif')==true
        ImageTiff = Tiff(filepath);
        pxsz = getTag(ImageTiff,'XResolution'); 
        Image = imread(filepath);
        units = getTag(ImageTiff,'ResolutionUnit');
        if isempty(units)
            warning('Unit of dm3 file is empty, Skip evaluation of %s\n',filepath)
            continue
        elseif units == 2
            %Unit is in dpi (inches) --> scale to nm/px
            pxsz = pxsz *2.54*10^(-7);
            pxsz = 1/pxsz; %nm/px
            units = 'nm';
            
        elseif units == 3 %px/cm
            %Unit is in cm --> scale to nm
            pxsz = pxsz *10^(-7);
            pxsz = 1/pxsz; %nm/px
            units = 'nm';
        else
            warning('Something is wrong with the unit\n')
        end
    end

    
    % apply filter
    Image=imgaussfilt(Image,2);
    % convert to uint8
    Image = uint8(255 * mat2gray(Image));
    % Segmentation
    tSegStart=toc
    C = semanticseg(Image, net);
    tSegEnd=toc
    fprintf('Segmentation of file No. %i completed\n',i)
    C =uint8(C);
    
    % Show image
%     figure
%     CB=labeloverlay(Image,C);
%     imshowpair(Image,CB,'montage')
%  
    %% get properties
    [Eval,partProp,B,Mlines] = particlePropertiesEval(C,2,pxsz,Image,...
        'Convexthresh',Convexthresh,'PSep',PSep,'minmarker',minmarker,...
        'removecount',removecount, 'Mode', Mode);

    if isempty(partProp)
        warning('Property measurement is empty\n No data generated\n')
        
        continue
    end

    % Make image Overlays
    B2=labeloverlay(Image,C,'Colormap',cmap);
    

    %% Add Scalebar to Image
    Image=scalebar(Image,pxsz,units);

    %% Save Results
    % make directory for results
    mkdir(resfolder,expName) 
    
    % save overlay image and raw image     
    imwrite(B,[resfolder '\' expName '\' expName '_ImageOverlay.png'])
    imwrite(B2,[resfolder '\' expName '\' expName '_ImageOverlayRaw.png'])
    imwrite(Image,[resfolder '\' expName '\' expName '_Image.png'])

    if PSep == true
        Mlines = imdilate(Mlines,strel('square',3));
         Mlines=labeloverlay(Image,Mlines,'Colormap',[1,1,1]);
        imwrite(Mlines,[resfolder '\' expName '\' expName 'ImageOverlayLines.png'])
    end

    % save results as Matlab file
    save([resfolder '\' expName '\EvalData_' expName '.mat'],'Eval','partProp','C','pxsz')

    if strcmp(saveas,'xlsx')==true
        % Write to Excel file
        writetable(partProp,[resfolder '\' expName '\EvalData_' expName '.xlsx'])
        xlswrite([resfolder '\' expName '\EvalData_' expName '.xlsx'],Eval,'MeanValues','B2')
        xlswrite([resfolder '\' expName '\EvalData_' expName '.xlsx'],{'mean Value','standard deviation'},'MeanValues','B1')
        rows={'Area [nm^2]'; 'Equivalent Diameter [nm]';...
            'min. Feret Diameter [nm]'; 'max. Feret Diameter [nm]';'Circularity [-]'};
        xlswrite([resfolder '\' expName '\EvalData_' expName '.xlsx'],rows,'MeanValues','A2')
        tend=toc
    elseif strcmp(saveas,'csv')==true
        writetable(partProp,[resfolder '\' expName '\EvalData_' expName '.csv'])
    end

if batch == 1  && allinone > 0
    AllpartProp = [AllpartProp; partProp];
end
end

if batch == 1 && allinone > 0
    writetable(AllpartProp,[resfolder '\' expName '_all_Images_Results.xlsx'])
end

File: PythonVersion/ANTEMA_main.py
""" Main script for Segmentation Analysis of HRTEM images
 Run this script to run the automated nanoparticle transmission electron
 micrograph analyzer (ANTEMA)


 Copyright (C) 2022  University of Duisburg-Essen
 
     This program is free software: you can redistribute it and/or modify
     it under the terms of the GNU General Public License as published by
     the Free Software Foundation, either version 3 of the License, or
     (at your option) any later version.
 
     This program is distributed in the hope that it will be useful,
     but WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
     GNU General Public License for more details.
 
     You should have received a copy of the GNU General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.
"""


import glob
import matplotlib.pyplot as plt
from ncempy.io import dm
import numpy as np
from scipy.ndimage import gaussian_filter
from scipy.stats import zscore
from skimage import morphology, measure
from tkinter import Tk, filedialog
import tensorflow as tf
from keras.utils import normalize
import os
import NPModel 





class ANTEMA:
    """ Main class for ANTEMA """

    def __init__(self,netname = 'NP-net-weights',Convexthresh = 0.95,PSep = True, saveas = 'xlsx', removecount = 0.5, minmarker = 0.5, SepMode = 'UECS'):
        """ Initialize with default parameters and add all inputs to self"""
        self.Convexthresh = Convexthresh
        self.PSep = PSep
        self.saveas = saveas
        self.removecount = removecount
        self.minmarker = minmarker
        self.SepMode = SepMode

    def batchmode(self):
        """Open file dialog to get the folder path of the folder containing all .dm3 files to be segmented and the paths to all the files"""
        root = Tk()
        root.attributes('-topmost',True)
        root.iconify()
        pathdir = filedialog.askdirectory(title = 'Select folder containing all .dm3 files to be segmented', parent = root)
        root.destroy()

        """Get all the files in the folder"""
        files = glob.glob(pathdir + '/*.dm3')
        """Ask How the files hould be saved, 0: one file for each image 1: all image data in one file 2: both"""
        self.saveas = int(input('How do you want to save the data?\n 0: one file for each image\n 1: all image data in one file\n 2: both'))
        """ Check if the input is 0,1 or 2 and if not repeat the question"""
        while self.saveas not in [0,1,2]:
            self.saveas = int(input('Input out of bounds. \nPlease type one of the following numbers:\n 0: one file for each image\n 1: all image data in one file\n 2: both'))
        for file in files:
            """Run the segmentation for each file"""
            self.run(file)
    
    def singlemode(self):
        """Open file dialog to get the path to the file to be segmented"""
        root = Tk()
        root.attributes('-topmost',True)
        root.iconify()
        path = filedialog.askopenfilename(title = 'Select dm3 file to be segmented', filetypes=[("dm3 files", "*.dm3")], parent = root)
        root.destroy()
        
        """Run the segmentation for the file"""
        self.run(path)
    
    def run(self, path):
        """Run the segmentation for the file"""
        
        """Load the dm3 file as a numpy array"""
        dm3 = dm.dmReader(path)
        image = dm3['data']
        pxSize = dm3['pixelSize'][0] 

        plt.figure()
        plt.imshow(image,cmap = 'gray')
        plt.show()
        """Apply a gauss filter and convert image to a uint8 rgb image"""
        image = gaussian_filter(image,2)
        '''normalize the image to values between 0 and 1 similarly to mat2gray in MATLAB'''
        image = self.mat2gray(image)
        image = (image*255).round().astype(np.uint8)

        '''apply zscore normalization with the values specified in the MATLAB code'''
        mean = 114.2725
        std =  58.9159
        image = (image - mean)/std
        image = np.stack((image,image,image),axis = 2)


        """Load the model"""
        model = NPModel.load_model([None,None,3], debug=True)

        '''Convert image to tensor and expand dimensions to fit the model input'''
        image = np.expand_dims(image, axis=0)
        image =  tf.convert_to_tensor(image, dtype=tf.float32)


        """Run the segmentation"""
        M = model.predict(image)
        
        M = np.squeeze(M)
        Mbackground = M[:,:,0]>0.5
        Mparticle = M[:,:,1]>0.5
        image = np.squeeze(image)
       
        exit()
        ### Code below is not finished and needs to be implemented
        """ Get particle properties"""
        props = self.particlePropertiesEval(Mparticle,pxSize,image )
        # Ab hier heavy testing!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

        """Make image overlay"""
        overlay = self.overlay(image, M) ### probably already exists

        """ Add a scalebar to the image"""
        image = self.scalebar(image,pxSize)

        """Save the data""" #### not finished, needs to be implemented
        if self.saveas == 0:
            """Save each image in a seperate file"""
            pass
        elif self.saveas == 1:
            """Save all images in one file"""
            pass
        elif self.saveas == 2:
            """Save each image in a seperate file and save all images in one file"""
            pass
            
    def scalebar(self,image,pxSize):
        """Add a scalebar to the image"""
        ### not yet immplemented
        pass

    def particlePropertiesEval(self,M,pxSize,image):
        """Measure the particle properties"""
        """Make M into a integer array"""
        # Not yet implemented
        M = M.astype(int)
               
        """Remove all Areas with an equivalence diameter of lower than the set removecount diameter nm size"""
        removecount = int(np.pi/4*(self.removecount/pxSize)**2)
        M = morphology.remove_small_objects(M,removecount)

        """If separation routine is activated, separate particles with convexity lower than the threshold"""
        if self.PSep:
            """ Measure the particle properties
            Measure the boundingbox, perimeter, convex image and euler number of each particle"""
            label_image = measure.label(M)
            props = measure.regionprops(label_image, property = ['bbox','perimeter','image','image_convex','euler_number'])
            ''' Measure the convexity of each particle by dividing the perimeter of the particle by the perimeter of the convex image'''
            for prop in props:
                prop['convexity'] = prop['perimeter']/measure.perimeter(prop['image_convex'])

                if prop['convexity'] < self.Convexthresh:
                    """"Check if Euler Number is below -10 and if it is it should be removed from the evaluation and the image"""
                    if prop['euler_number'] < -10:
                        M[prop['image']] = 0 # CHECK IF THIS WORKS
                        """Give the user a warning message that the particle was removed from the image due to too many holes"""
                        Warning('Particle with too many holes removed from the image')
                    """Check which separation method to use"""
                    if self.SepMode == 'UECS':
                        Msplit = self.ParticleSeparationUECS(prop['image'],pxSize)
                        """Remove the old particle from the image and add the new particles as layers in the stack"""
                    elif self.SepMode == 'Watershed':
                        Msplit = self.ParticleSeparationWatershed(prop['image'],pxSize)
                        """Remove the old particle from the image and add the new particles as layers in the stack"""
        
        else:
            """Get the particle properties (area, equivalent diameter, perimeter, minimum and maximum feret diameter, boundinb box, centroid position and convexity )"""
            label_image = measure.label(M)
            props = measure.regionprops_table(label_image, property = ['area','perimeter','minor_axis_length','major_axis_length','bbox','centroid'])


        

    def ParticleSeparationUECS(self, M, pxSize):
        """Separate particles that are connected by a small bridge"""
        # Not yet implemented
        minmarker = int(np.pi/4*(self.minmarker/pxSize)**2)

        updated = True
        iter_count = 0
        while updated:
            updated = False
            label_image = measure.label(M)
            """Get convex area, area, perimeter, convex image and minor axis length of each particle"""
            props = measure.regionprops(label_image, property = ['area','perimeter','image','image_convex','minor_axis_length'])
            if np.mod(iter_count,2) == 1:
                se = morphology.disk(1)
            else:
                """ se2    = strel('arbitrary', ones(2,2));"""
                se = morphology.arbitary(2) # CHECK IF THIS WORKS
            
            for prop in props:


                iter_count += 1



        return Msplit
        pass

    def ParticleSeparationWatershed(self, M, pxSize):
        """Separate particles using the watershed algorithm"""
        ## Not yet implemented
        pass

    def mat2gray(self,image):
        """Normalize the image to values between 0 and 1 similarly to mat2gray in MATLAB"""
        image = (image - np.min(image))/(np.max(image)-np.min(image))
        return image

A =ANTEMA()
A.singlemode()

        









        

        




        




    



